### ISSUE TITLE
:lua* do not support [range] despite stated in documentation

### ISSUE THREAD

**Issue by ZyX-I** [neovim/neovim#7177]
Currently the only command actually supporting `[range]` is `luado`, but documentation (copied from Vim) states that `:lua` and `:luafile` also should, by storing the range in `vim.firstline` and `vim.lastline`. In the current state other commands ignore range completely.

Regarding the implementation I currently more like the idea of adding a metatable to `vim` “module” which will do the following two things:

1. Make `vim` module read-only.
2. Provide firstline and lastline.

This would protect vim module from accidental changes (and similar metatable is to be applied to `vim.types` and `vim.api`; “similar” = “read-only, but no `*line`”). It would still be not impossible to modify our tables, metatable can only protect from some accidental modifications.

**Comment by ZyX-I** [comment]
Created the issue as I don’t have time to do the change right now.

**Comment by justinmk** [comment]
# Status

- [#27167](https://github.com/neovim/neovim/pull/27167) implements `:{range}lua` so that Lua snippets can be executed by selecting it.
    - So `vim.firstline`, `vim.lastline` don't make sense (and are out of scope) for `:lua`.
- `vim.firstline`, `vim.lastline` still would be useful for `:luado` and `:luafile`.

**Comment by zeertzjq** [comment]
`vim.firstline` and `vim.lastline` are indeed less useful for `:lua` because of the existence of Lua user commands.

**Comment by lewis6991** [comment]
Surprised there's not a way to do this in vimscript already.

My instinct would be to add something like `vim.fn.line('r0')` and `vim.fn.line('r$')`.

**Comment by zeertzjq** [comment]
> Surprised there's not a way to do this in vimscript already.
> 
> My instinct would be to add something like `vim.fn.line('r0')` and `vim.fn.line('r$')`.

This doesn't not make sense for Vim script as every line in a Vim script is an Ex command.

**Comment by lewis6991** [comment]
> This doesn't not make sense for Vim script as every line in a Vim script is an Ex command, so if line() is called in Vim script it's called in an Ex command.

`vim.fn.line()` does not invoke any ex command.

**Comment by zeertzjq** [comment]
> > This doesn't not make sense for Vim script as every line in a Vim script is an Ex command, so if line() is called in Vim script it's called in an Ex command.
> 
> `vim.fn.line()` does not invoke any ex command.

Yes, then what is the point of having to go through typval conversion for something that doesn't really apply to Vim script?

**Comment by lewis6991** [comment]
Because it's a standard interface that people are familiar with and the typval conversion isn't really an issue with those kind of functions.

I wouldn't be against adding `vim.line()` and `vim.getpos()` though.

**Comment by lewis6991** [comment]
Just realised this is only for the `:lua` which is quite a limited scope. Maybe we can just set `firstline` and `lastline` as locals to the code block. E.g. running a `:lua <code>` would be equivalent to:

```lua
:lua (function(firstline, lastline) <code> end)()
```

**Comment by justinmk** [comment]
> set `firstline` and `lastline` as locals to the code block

What are the downsides of using the same `vim.firstline`/`vim.lastline` names? Advantage is that we avoid some unnecessary differences. The scope can still be limited to :lua.

**Comment by lewis6991** [comment]
> Advantage is that we avoid some unnecessary differences.

What differences?

Doing dynamic stuff in `vim.*` like this feels weird, and breaks the model that `vim.*` is an entry point into the API.

And tbh, is this something we even need to do? And value of `:lua` is significantly reduced compared to the past. If someone really wants `firstline` and `lastline`, then can create a command via `nvim_create_user_command` and use that, and that will further give them all the invocation metadata available to ex commands.

**Comment by justinmk** [comment]
> value of `:lua` is significantly reduced compared to the past

I use it all the time, it's a great way to run code from help files (and we should do that for more kinds of code). 

> create a command via `nvim_create_user_command` and use

I haven't missed firstline/lastline much, but that might be because I am used to `:global`. But the idea is to have a Lua alternative to :global and substitute.

This is an easy feature to add and if it introduces new use-cases for using Lua interactively, that's good. Same reason we added `:=` and `:lua =`, even though those could be achieved with custom commands.

**Comment by lewis6991** [comment]
> The names

Hmm you might have misunderstood my suggestion. I meant we literally set the local names `firstline` and `lastline` within the luachunk, they can be accessed as upvals and can only be accessed via that chunk.

If we add `vim.firstline` then every function within the call tree of the chunk can access it.

**Comment by justinmk** [comment]
If we're changing stuff, instead of multiple magic local names, WDYT about a single `ev` magic local which may contain other fields? Also could help with code completion (`ev.<tab>`).
